File: snips-nlu-ontology/src/ontology.rs
use std::ops::Range;

#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq)]
pub struct IntentParserResult {
    pub input: String,
    pub intent: Option<IntentClassifierResult>,
    pub slots: Option<Vec<Slot>>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct IntentClassifierResult {
    pub intent_name: String,
    pub probability: f32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Slot {
    pub raw_value: String,
    pub value: SlotValue,
    pub range: Option<Range<usize>>,
    pub entity: String,
    pub slot_name: String,
}

impl Slot {
    pub fn new_custom(
        value: String,
        range: Range<usize>,
        entity: String,
        slot_name: String,
    ) -> Slot {
        Slot {
            raw_value: value.clone(),
            value: SlotValue::Custom(value.into()),
            range: Some(range),
            entity,
            slot_name,
        }
    }
}

impl Slot {
    pub fn with_slot_value(self, slot_value: SlotValue) -> Slot {
        Slot {
            raw_value: self.raw_value,
            value: slot_value,
            range: self.range,
            entity: self.entity,
            slot_name: self.slot_name,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum SlotValue {
    Custom(StringValue),
    Number(NumberValue),
    Ordinal(OrdinalValue),
    Percentage(PercentageValue),
    InstantTime(InstantTimeValue),
    TimeInterval(TimeIntervalValue),
    AmountOfMoney(AmountOfMoneyValue),
    Temperature(TemperatureValue),
    Duration(DurationValue),
    MusicArtist(StringValue)
}

/// This struct is required in order to use serde Internally tagged enum representation
#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct StringValue {
    pub value: String,
}

impl From<String> for StringValue {
    fn from(string: String) -> Self {
        StringValue { value: string }
    }
}

impl From<&'static str> for StringValue {
    fn from(str: &str) -> Self {
        StringValue {
            value: str.to_string(),
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub struct NumberValue {
    pub value: f64,
}

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub struct OrdinalValue {
    pub value: i64,
}

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub struct PercentageValue {
    pub value: f64,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct InstantTimeValue {
    pub value: String,
    pub grain: Grain,
    pub precision: Precision,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct TimeIntervalValue {
    pub from: Option<String>,
    pub to: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct AmountOfMoneyValue {
    pub value: f32,
    pub precision: Precision,
    pub unit: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct TemperatureValue {
    pub value: f32,
    pub unit: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
pub struct DurationValue {
    pub years: i64,
    pub quarters: i64,
    pub months: i64,
    pub weeks: i64,
    pub days: i64,
    pub hours: i64,
    pub minutes: i64,
    pub seconds: i64,
    pub precision: Precision,
}

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub enum Grain {
    Year = 0,
    Quarter = 1,
    Month = 2,
    Week = 3,
    Day = 4,
    Hour = 5,
    Minute = 6,
    Second = 7,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Copy, Clone)]
pub enum Precision {
    Approximate,
    Exact,
}

#[cfg(test)]
mod tests {
    use serde_json;

    use super::*;

    #[test]
    fn test_custom_slot() {
        let slot = Slot {
            raw_value: "value".into(),
            value: SlotValue::Custom("value".into()),
            range: None,
            entity: "toto".into(),
            slot_name: "toto".into(),
        };
        assert!(serde_json::to_string(&slot).is_ok());
        assert!(serde_json::from_str::<Slot>(&serde_json::to_string(&slot).unwrap()).is_ok());
    }

    #[test]
    fn test_builtin_slot_1() {
        let slot = Slot {
            raw_value: "fifth".into(),
            value: SlotValue::Ordinal(OrdinalValue { value: 5 }),
            range: None,
            entity: "toto".into(),
            slot_name: "toto".into(),
        };
        assert!(serde_json::to_string(&slot).is_ok());
        assert!(serde_json::from_str::<Slot>(&serde_json::to_string(&slot).unwrap()).is_ok());
    }

    #[test]
    fn test_builtin_slot_2() {
        let slot = Slot {
            raw_value: "some_value".into(),
            value: SlotValue::InstantTime(InstantTimeValue {
                value: "some_value".into(),
                grain: Grain::Year,
                precision: Precision::Exact,
            }),
            range: None,
            entity: "toto".into(),
            slot_name: "toto".into(),
        };
        assert!(serde_json::to_string(&slot).is_ok());
        assert!(serde_json::from_str::<Slot>(&serde_json::to_string(&slot).unwrap()).is_ok());
    }
}

File: platforms/kotlin/src/main/kotlin/ai/snips/nlu/ontology/Ontology.kt
package ai.snips.nlu.ontology

import ai.snips.nlu.ontology.SlotValue.AmountOfMoneyValue
import ai.snips.nlu.ontology.SlotValue.CityValue
import ai.snips.nlu.ontology.SlotValue.CountryValue
import ai.snips.nlu.ontology.SlotValue.CustomValue
import ai.snips.nlu.ontology.SlotValue.DurationValue
import ai.snips.nlu.ontology.SlotValue.InstantTimeValue
import ai.snips.nlu.ontology.SlotValue.MusicAlbumValue
import ai.snips.nlu.ontology.SlotValue.MusicArtistValue
import ai.snips.nlu.ontology.SlotValue.MusicTrackValue
import ai.snips.nlu.ontology.SlotValue.NumberValue
import ai.snips.nlu.ontology.SlotValue.OrdinalValue
import ai.snips.nlu.ontology.SlotValue.PercentageValue
import ai.snips.nlu.ontology.SlotValue.RegionValue
import ai.snips.nlu.ontology.SlotValue.TemperatureValue
import ai.snips.nlu.ontology.SlotValue.TimeIntervalValue
import ai.snips.nlu.ontology.SlotValue.Type.AMOUNT_OF_MONEY
import ai.snips.nlu.ontology.SlotValue.Type.CITY
import ai.snips.nlu.ontology.SlotValue.Type.COUNTRY
import ai.snips.nlu.ontology.SlotValue.Type.CUSTOM
import ai.snips.nlu.ontology.SlotValue.Type.DURATION
import ai.snips.nlu.ontology.SlotValue.Type.INSTANT_TIME
import ai.snips.nlu.ontology.SlotValue.Type.MUSICALBUM
import ai.snips.nlu.ontology.SlotValue.Type.MUSICARTIST
import ai.snips.nlu.ontology.SlotValue.Type.MUSICTRACK
import ai.snips.nlu.ontology.SlotValue.Type.NUMBER
import ai.snips.nlu.ontology.SlotValue.Type.ORDINAL
import ai.snips.nlu.ontology.SlotValue.Type.PERCENTAGE
import ai.snips.nlu.ontology.SlotValue.Type.REGION
import ai.snips.nlu.ontology.SlotValue.Type.TEMPERATURE
import ai.snips.nlu.ontology.SlotValue.Type.TIME_INTERVAL
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.JsonTypeInfo
import org.parceler.Parcel
import org.parceler.Parcel.Serialization.BEAN
import org.parceler.ParcelConstructor
import org.parceler.ParcelProperty

@Parcel(BEAN)
data class Range @ParcelConstructor constructor(@ParcelProperty("start") val start: Int,
                                                @ParcelProperty("end") val end: Int)

@Parcel(BEAN)
data class Slot @ParcelConstructor constructor(@ParcelProperty("rawValue") val rawValue: String,
                                               @ParcelProperty("value") val value: SlotValue,
                                               //use a MutableList here to make parceler happy
                                               @ParcelProperty("alternatives") val alternatives: MutableList<SlotValue>,
                                               @ParcelProperty("range") val range: Range,
                                               @ParcelProperty("entity") val entity: String,
                                               @ParcelProperty("slotName") val slotName: String,
                                               @ParcelProperty("confidenceScore") val confidenceScore: Float?)

enum class Precision { APPROXIMATE, EXACT }

enum class Grain { YEAR, QUARTER, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND }

// TODO : add converters to JSR310 / ThreeTen types
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "kind")
@JsonSubTypes(
        Type(value = CustomValue::class, name = "Custom"),
        Type(value = NumberValue::class, name = "Number"),
        Type(value = OrdinalValue::class, name = "Ordinal"),
        Type(value = InstantTimeValue::class, name = "InstantTime"),
        Type(value = TimeIntervalValue::class, name = "TimeInterval"),
        Type(value = AmountOfMoneyValue::class, name = "AmountOfMoney"),
        Type(value = TemperatureValue::class, name = "Temperature"),
        Type(value = DurationValue::class, name = "Duration"),
        Type(value = PercentageValue::class, name = "Percentage"),
        Type(value = MusicAlbumValue::class, name = "MusicAlbum"),
        Type(value = MusicArtistValue::class, name = "MusicArtist"),
        Type(value = MusicTrackValue::class, name = "MusicTrack"),
        Type(value = CityValue::class, name = "City"),
        Type(value = CountryValue::class, name = "Country"),
        Type(value = RegionValue::class, name = "Region")
)
sealed class SlotValue(val kind: Type) {

    @JsonIgnore
    @Deprecated("use kind instead")
    val type = kind

    @Parcel
    enum class Type {
        @JsonProperty("Custom") CUSTOM,
        @JsonProperty("Number") NUMBER,
        @JsonProperty("Ordinal") ORDINAL,
        @JsonProperty("InstantTime") INSTANT_TIME,
        @JsonProperty("TimeInterval") TIME_INTERVAL,
        @JsonProperty("AmountOfMoney") AMOUNT_OF_MONEY,
        @JsonProperty("Temperature") TEMPERATURE,
        @JsonProperty("Duration") DURATION,
        @JsonProperty("Percentage") PERCENTAGE,
        @JsonProperty("MusicAlbum") MUSICALBUM,
        @JsonProperty("MusicArtist") MUSICARTIST,
        @JsonProperty("MusicTrack") MUSICTRACK,
        @JsonProperty("City") CITY,
        @JsonProperty("Country") COUNTRY,
        @JsonProperty("Region") REGION
    }

    @Parcel(BEAN)
    data class CustomValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(CUSTOM)

    @Parcel(BEAN)
    data class NumberValue @ParcelConstructor constructor(@ParcelProperty("value") val value: Double) : SlotValue(NUMBER)

    @Parcel(BEAN)
    data class PercentageValue @ParcelConstructor constructor(@ParcelProperty("value") val value: Double) : SlotValue(PERCENTAGE)

    @Parcel(BEAN)
    data class OrdinalValue @ParcelConstructor constructor(@ParcelProperty("value") val value: Long) : SlotValue(ORDINAL)

    @Parcel(BEAN)
    data class InstantTimeValue @ParcelConstructor constructor(
            @ParcelProperty("value") val value: String,
            @ParcelProperty("grain") val grain: Grain,
            @ParcelProperty("precision") val precision: Precision) : SlotValue(INSTANT_TIME)

    @Parcel(BEAN)
    data class TimeIntervalValue @ParcelConstructor constructor(
            @ParcelProperty("from") val from: String?,
            @ParcelProperty("to") val to: String?) : SlotValue(TIME_INTERVAL)

    @Parcel(BEAN)
    data class AmountOfMoneyValue @ParcelConstructor constructor(
            @ParcelProperty("value") val value: Float,
            @ParcelProperty("precision") val precision: Precision,
            @ParcelProperty("unit") val unit: String?) : SlotValue(AMOUNT_OF_MONEY)

    @Parcel(BEAN)
    data class TemperatureValue @ParcelConstructor constructor(
            @ParcelProperty("value") val value: Float,
            @ParcelProperty("unit") val unit: String?) : SlotValue(TEMPERATURE)

    @Parcel(BEAN)
    data class DurationValue @ParcelConstructor constructor(
            @ParcelProperty("years") val years: Long,
            @ParcelProperty("quarters") val quarters: Long,
            @ParcelProperty("months") val months: Long,
            @ParcelProperty("weeks") val weeks: Long,
            @ParcelProperty("days") val days: Long,
            @ParcelProperty("hours") val hours: Long,
            @ParcelProperty("minutes") val minutes: Long,
            @ParcelProperty("seconds") val seconds: Long,
            @ParcelProperty("precision") val precision: Precision) : SlotValue(DURATION)

    @Parcel(BEAN)
    data class MusicAlbumValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(MUSICALBUM)

    @Parcel(BEAN)
    data class MusicArtistValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(MUSICARTIST)

    @Parcel(BEAN)
    data class MusicTrackValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(MUSICTRACK)

    @Parcel(BEAN)
    data class CityValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(CITY)

    @Parcel(BEAN)
    data class CountryValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(COUNTRY)

    @Parcel(BEAN)
    data class RegionValue @ParcelConstructor constructor(@ParcelProperty("value") val value: String) : SlotValue(REGION)
}

@Parcel(BEAN)
data class IntentClassifierResult @ParcelConstructor constructor(
        @ParcelProperty("intentName") val intentName: String?,
        @ParcelProperty("confidenceScore") val confidenceScore: Float)

@Parcel(BEAN)
data class IntentParserAlternative @ParcelConstructor constructor(
        @ParcelProperty("intent") val intent: IntentClassifierResult,
        @ParcelProperty("slots") val slots: List<Slot>)

@Parcel(BEAN)
data class IntentParserResult @ParcelConstructor constructor(
        @ParcelProperty("input") val input: String,
        @ParcelProperty("intent") val intent: IntentClassifierResult,
        @ParcelProperty("slots") val slots: List<Slot>,
        @ParcelProperty("alternatives") val alternatives: List<IntentParserAlternative>)
